core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala

```diff
@@ -117,26 +117,32 @@ class BrokerMetadataListener(
       } finally {
         reader.close()
       }
-      _publisher.foreach(publish)
 
-      // If we detected a change in metadata.version, generate a local snapshot
-      val metadataVersionChanged = Option(_delta.featuresDelta()).exists { featuresDelta =>
-        featuresDelta.metadataVersionChange().isPresent
+      _bytesSinceLastSnapshot = _bytesSinceLastSnapshot + results.numBytes
+      if (shouldSnapshot()) {
+        maybeStartSnapshot()
       }
 
-      snapshotter.foreach { snapshotter =>
-        _bytesSinceLastSnapshot = _bytesSinceLastSnapshot + results.numBytes
-        if (shouldSnapshot() || metadataVersionChanged) {
-          if (snapshotter.maybeStartSnapshot(_highestTimestamp, _delta.apply())) {
-            _bytesSinceLastSnapshot = 0L
-          }
-        }
-      }
+      _publisher.foreach(publish)
     }
   }
 
   private def shouldSnapshot(): Boolean = {
-    _bytesSinceLastSnapshot >= maxBytesBetweenSnapshots
+    (_bytesSinceLastSnapshot >= maxBytesBetweenSnapshots) || metadataVersionChanged()
+  }
+
+  private def metadataVersionChanged(): Boolean = {
+    _publisher.nonEmpty && Option(_delta.featuresDelta()).exists { featuresDelta =>
```

Comment:
Do you mind writing a comment as to why we check that the publisher is set. If I understand this correctly it is not a correctness issue but a performance issue, right? If I remember correctly, @mumrah mentioned that he wanted to generate a snapshot whenever the metadata version changes. Unfortunately, I couldn't find a mention of this in KIP-778.

With this change this is no longer true. What do you think @mumrah ?